packages/server/graphql/mutations/archiveTimelineEvent.ts

```diff
@@ -0,0 +1,71 @@
+import {GraphQLID, GraphQLNonNull} from 'graphql'
+import getRethink from '../../database/rethinkDriver'
+import {GQLContext} from '../graphql'
+import ArchiveTimelineEventPayload from '../types/ArchiveTimelineEventPayload'
+import publish from '../../utils/publish'
+import {SubscriptionChannel} from 'parabol-client/types/constEnums'
+import TimelineEvent from '../../database/types/TimelineEvent'
+import {getUserId, isTeamMember} from '../../utils/authorization'
+import standardError from '../../utils/standardError'
+
+const archiveTimelineEvent = {
+  type: GraphQLNonNull(ArchiveTimelineEventPayload),
+  description: `Archive a timeline event`,
+  args: {
+    timelineEventId: {
+      type: GraphQLNonNull(GraphQLID),
+      description: 'the id for the timeline event'
+    },
+    meetingId: {
+      type: GraphQLNonNull(GraphQLID),
+      description: 'the meeting id for the timeline event'
+    },
+    teamId: {
+      type: GraphQLNonNull(GraphQLID),
+      description: 'the team id for the timeline event'
+    }
+  },
+  resolve: async (
+    _source,
+    {timelineEventId, meetingId, teamId},
+    {authToken, dataLoader, socketId: mutatorId}: GQLContext
+  ) => {
+    const r = await getRethink()
+    const operationId = dataLoader.share()
+    const subOptions = {mutatorId, operationId}
+    const viewerId = getUserId(authToken)
+
+    // AUTH
+    if (!isTeamMember(authToken, teamId)) {
+      return standardError(new Error('Team not found'), {userId: viewerId})
+    }
+
+    // VALIDATION
+    const timelineEvents = await dataLoader.get('timelineEventsByMeetingId').load(meetingId)
+    if (!timelineEvents) {
+      return {error: {message: 'TimelineEvents not found'}}
+    }
+
+    // RESOLUTION
+    await Promise.all(
+      timelineEvents.map((timelineEvent: TimelineEvent) => {
+        timelineEvent.isActive = false
+        r.table('TimelineEvent')
+          .get(timelineEvent.id)
+          .update({
+            isActive: false
+          })
+          .run()
+      })
+    )
+
+    timelineEvents.map(({id: timelineEventId}) => {
+      const data = {timelineEventId}
+      publish(SubscriptionChannel.TEAM, teamId, 'ArchiveTimelineEventSuccess', data, subOptions)
```

Comment:
As written this is O(N^2). If a team has 10 members, and each member has their own timeline event, then it'll publish every eventId to every member for 10 * 10 = 100 messages.

We can make this run in O(N) time by using the `NOTIFICATION` channel, since that channel only has 1 listener (the userId).

Reply:
I’m glad you grok big O 😆